src/bigint/division.rs

```diff
@@ -446,3 +446,51 @@ impl CheckedDiv for BigInt {
         Some(self.div(v))
     }
 }
+
+impl CheckedEuclid for BigInt {
+    #[inline]
+    fn checked_div_euclid(&self, v: &BigInt) -> Option<BigInt> {
+        if v.is_zero() {
+            return None;
+        }
+        Some(self.div_euclid(v))
+    }
+
+    #[inline]
+    fn checked_rem_euclid(&self, v: &BigInt) -> Option<BigInt> {
+        if v.is_zero() {
+            return None;
+        }
+        Some(self.rem_euclid(v))
+    }
+}
+
+impl Euclid for BigInt {
+    #[inline]
+    fn div_euclid(&self, v: &BigInt) -> BigInt {
+        let q = self / v;
+        if self % v < BigInt::zero() {
```

Comment:
This should use `div_rem` so we don't have to divide twice.

Also, `is_negative()` and `is_positive()` would be more direct than all the comparisons to `zero()`.